STYLEGUIDE.md

```diff
@@ -0,0 +1,367 @@
+
+### Format Contributions
+
+### Prefer [Rule]
+
+#### Why
+
+_Description_
+
+#### Prefer
+
+_Example_
+
+#### Avoid
+
+_Example_
+
+---
+
+## <a id="Typing">Typing</a>
+
+### Prefer constants and types over enums.
+
+#### Why
+
+Ensures that TypeScript doesn't prevent people from using the values directly rather than consuming the enum.
+
+#### Prefer
+
+```typescript
+export const someConstant = {
+  optionOne = 'option 1',
+  optionTwo = 'option 2',
+} as const;
+```
+
+#### Avoid
+
+```typescript
+enum someConstant = {
+ optionOne: 'option 1',
+ optionTwo: 'option 2',
+}
+```
+
+---
+
+### When creating TS entities, place them at the top of the file, above JS
+
+---
+
+## <a id="Functions">Functions</a>
+
+### Prefer ternaries over simple `if/else` blocks
+
+#### Why
+
+Keeps code cleaner and easier to read.
+
+#### Prefer
+
+```typescript
+return isBlah ? a : b;
+```
+
+#### Avoid
+
+```typescript
+if (isBlah) {
+  return a;
+} else {
+  return b;
+}
+```
+
+---
+
+### All new (exported) functions should have a JSDoc comment explaining functionality.
+
+#### Why
+
+Documenting functions with [JSDoc](https://jsdoc.app/about-getting-started.html) helps PR reviewers better understand the intent of the function and also documents its use for other engineers using the code base. Better yet, [VS Code will automatically show in-line descriptions of a function](https://code.visualstudio.com/docs/languages/javascript#_jsdoc-support) when a user invokes that function in their editor.
+
+JSDocs should also clearly mention any major assumptions made about the function's input and output that may not be immediately obvious to an engineer unfamiliar with this part of the code base.
+
+#### Prefer
+
+```typescript
+/**
+ * Get the next applicable instance for upgrade. The function handles regions and cloud providers
+ * but does NOT work for Serverless clusters
+ * @param clusterDescription the backbone cluster description model
+ * @param providerOptionsModel the backbone provider options model
+ * @returns the next upgrade instance for this cluster
+ */
+export const getNextUpgradeInstance = (
+  clusterDescription: ClusterDescription,
+  providerOptionsModel: ProviderOptions,
+) => {
+  // ...
+};
+```
+
+#### Avoid
+
+```typescript
+export const getNextUpgradeInstance = (
+  clusterDescription: ClusterDescription,
+  providerOptionsModel: ProviderOptions,
+) => {
+  // ...
+};
+```
+
+---
+
+### Prefer creating the function in the highest scope, and passing in all required data.
+
+#### Why
+
+Keeps functions pure.
+
+---
+
+## <a id="Variable-naming">Variable Naming</a>
+
+### Acronyms are considered a single entity in camel-casing
+
+#### Why
+
+Some treat each letter in an acronym as its own entity when an acronyme is included in a var name. We'll instead consider the acronym as a single entity, and only capitalize the first letter of it (if that).
```

Comment:
This is opposite of most React variables/types though (e.g. `type React.HTMLElement`)

Reply:
Would you prefer WIP? I can't think of another example where we don't do that other than HTML

Reply:
Hmm, I've never thought about this. It is a little easier to read when the entire acronym is capitalized. Are there examples of how other style guides handle this?

Reply:
I took this from Atlas' style guide!

Reply:
Most built-in examples use all caps (`CSS`, `XML`, `HTML`, `SVG`, `JSX`,  `JSON`, )
Though one example that's not even internally consistent is `XMLTHttpRequest`

Reply:
Kk, will leave this out

Reply:
I think that's fair. Not sure how much we use acronyms really (the only one that comes to mind is `FedRAMP`)